storage/src/storage/sqlite/sync.rs

```diff
@@ -0,0 +1,923 @@
+                block_order INTEGER NOT NULL
+            );
+            CREATE UNIQUE INDEX transaction_block_ordering ON transaction_blocks(block_id, block_order);
+            CREATE INDEX transaction_block_looukp ON transaction_blocks(transaction_id);
+            ")?;
+        }
+
+        if from <= 1 {
+            self.conn.execute_batch(
+                r"
+            CREATE TABLE IF NOT EXISTS miner_records(
+                id INTEGER PRIMARY KEY,
+                owner TEXT NOT NULL,
+                is_dummy INTEGER NOT NULL,
+                value INTEGER NOT NULL,
+                payload BLOB NOT NULL,
+                birth_program_id BLOB NOT NULL,
+                death_program_id BLOB NOT NULL,
+                serial_number_nonce BLOB NOT NULL,
+                commitment BLOB NOT NULL,
+                commitment_randomness BLOB NOT NULL
+            );
+            CREATE INDEX record_owner_lookup ON miner_records(owner);
+            CREATE INDEX record_commitment_lookup ON miner_records(commitment);
+            ",
+            )?;
+        }
+
+        if from == 0 {
+            self.conn
+                .execute("INSERT INTO migration VALUES (?)", [Self::SCHEMA_INDEX])?;
+        } else {
+            self.conn
+                .execute("UPDATE migration SET schema_index = ?", [Self::SCHEMA_INDEX])?;
+        }
+        Ok(())
+    }
+}
+
+impl SyncStorage for SqliteStorage {
+    fn init(&mut self) -> Result<()> {
+        self.conn.execute(
+            r"
+        CREATE TABLE IF NOT EXISTS migration(
+            schema_index INTEGER PRIMARY KEY
+        );",
+            [],
+        )?;
+        let current_index = self
+            .conn
+            .query_row(
+                r"
+        SELECT schema_index FROM migration LIMIT 1",
+                [],
+                |row| row.get(0),
+            )
+            .optional()?
+            .unwrap_or(0u32);
+        if current_index < Self::SCHEMA_INDEX {
+            self.migrate(current_index)?;
+        }
+        Ok(())
+    }
+
+    fn insert_block(&mut self, block: &SerialBlock) -> Result<()> {
+        let hash = block.header.hash();
+        let mut block_query = self.conn.prepare_cached(
+            r"
+        INSERT INTO blocks (hash,
+            previous_block_id,
+            previous_block_hash,
+            merkle_root_hash,
+            pedersen_merkle_root_hash,
+            proof,
+            time,
+            difficulty_target,
+            nonce)
+            VALUES (
+                ?,
+                (SELECT id from blocks where hash = ?),
+                ?,
+                ?,
+                ?,
+                ?,
+                ?,
+                ?,
+                ?
+            )
+        ",
+        )?;
+        block_query.execute::<&[&dyn ToSql]>(&[
+            &hash,
+            &block.header.previous_block_hash,
+            &block.header.previous_block_hash,
```

Comment:
should the `previous_block_hash` be inserted twice here?

Reply:
yes, we pass the parameter in twice. one is for a lookup subquery IIRC

Reply:
ah yes, I see it now :+1:; need to get used to this notation